src/toxcore/onion/client/mod.rs

```diff
@@ -2301,4 +2348,64 @@ mod tests {
 
         assert!(state.paths_pool.path_nodes.rand().is_some());
     }
+
+    #[test]
+    fn send_onion_request_udp() {
+        let (dht_pk, dht_sk) = gen_keypair();
+        let (real_pk, real_sk) = gen_keypair();
+        let (udp_tx, udp_rx) = mpsc::channel(1);
+        let (tcp_incoming_tx, _tcp_incoming_rx) = mpsc::unbounded();
+        let dht = DhtServer::new(udp_tx, dht_pk, dht_sk.clone());
+        let tcp_connections = TcpConnections::new(dht_pk, dht_sk, tcp_incoming_tx);
+        let onion_client = OnionClient::new(dht, tcp_connections, real_sk, real_pk);
+
+        let inner_onion_request = InnerOnionRequest::InnerOnionAnnounceRequest(InnerOnionAnnounceRequest {
+            nonce: gen_nonce(),
+            pk: gen_keypair().0,
+            payload: vec![42; 123],
+        });
+
+        let path = OnionPath::new([
+            PackedNode::new("127.0.0.1:12346".parse().unwrap(), &gen_keypair().0),
+            PackedNode::new("127.0.0.1:12347".parse().unwrap(), &gen_keypair().0),
+            PackedNode::new("127.0.0.1:12348".parse().unwrap(), &gen_keypair().0),
+        ], /* TCP */ false);
+        let saddr = "127.0.0.1:12345".parse().unwrap();
+        onion_client.send_onion_request(path, inner_onion_request, saddr).wait().unwrap();
+
+        let (_received, _udp_rx) = udp_rx.into_future().wait().unwrap();
+    }
+
+    #[test]
+    fn send_onion_request_tcp() {
+        let (dht_pk, dht_sk) = gen_keypair();
+        let (real_pk, real_sk) = gen_keypair();
+        let (udp_tx, udp_rx) = mpsc::channel(1);
+        let (tcp_incoming_tx, _tcp_incoming_rx) = mpsc::unbounded();
+        let dht = DhtServer::new(udp_tx, dht_pk, dht_sk.clone());
+        let tcp_connections = TcpConnections::new(dht_pk, dht_sk, tcp_incoming_tx);
+        let (_relay_incoming_rx, relay_outgoing_rx, relay_pk) = tcp_connections.add_client();
+        let onion_client = OnionClient::new(dht, tcp_connections, real_sk, real_pk);
+
+        let inner_onion_request = InnerOnionRequest::InnerOnionAnnounceRequest(InnerOnionAnnounceRequest {
+            nonce: gen_nonce(),
+            pk: gen_keypair().0,
+            payload: vec![42; 123],
+        });
+
+        let path = OnionPath::new([
+            PackedNode::new("127.0.0.1:12346".parse().unwrap(), &relay_pk),
+            PackedNode::new("127.0.0.1:12347".parse().unwrap(), &gen_keypair().0),
+            PackedNode::new("127.0.0.1:12348".parse().unwrap(), &gen_keypair().0),
+        ], /* TCP */ true);
```

Comment:
Please add TODO to replace bool with enum

Reply:
Or a new issue

Reply:
Simply added enum :)